crates/flux-fhir-analysis/src/conv.rs

```diff
@@ -349,33 +350,66 @@ impl<'a, 'tcx> ConvCtxt<'a, 'tcx> {
         bounded_ty: &rty::Ty,
         trait_id: DefId,
         args: &[fhir::GenericArg],
+        params: &[fhir::GenericParam],
         clauses: &mut Vec<rty::Clause>,
     ) -> QueryResult<()> {
         let mut into = vec![rty::GenericArg::Ty(bounded_ty.clone())];
         self.conv_generic_args_into(env, args, &mut into)?;
         self.fill_generic_args_defaults(trait_id, &mut into)?;
         let trait_ref = rty::TraitRef { def_id: trait_id, args: into.into() };
         let pred = rty::TraitPredicate { trait_ref };
-        clauses.push(rty::Clause::new(rty::ClauseKind::Trait(pred)));
+        let vars = params
+            .iter()
+            .map(|param| self.conv_trait_bound_generic_param(param))
+            .try_collect_vec()?;
+        clauses.push(rty::Clause::new(List::from_vec(vars), rty::ClauseKind::Trait(pred)));
         Ok(())
     }
 
+    fn conv_trait_bound_generic_param(
+        &self,
+        param: &fhir::GenericParam,
+    ) -> QueryResult<rty::BoundVariableKind> {
+        match &param.kind {
+            fhir::GenericParamKind::Lifetime => {
+                let def_id = param.def_id;
+                let name = self
+                    .genv
+                    .tcx
+                    .hir()
+                    .name(self.genv.tcx.hir().local_def_id_to_hir_id(def_id));
+                let kind = rty::BoundRegionKind::BrNamed(def_id.to_def_id(), name);
+                Ok(rty::BoundVariableKind::Region(kind))
+            }
+            fhir::GenericParamKind::Type { default: _ } => bug!("unexpected!"),
+            fhir::GenericParamKind::BaseTy => bug!("unexpected!"),
+        }
+    }
+
     fn conv_fn_bound(
         &self,
         env: &mut Env,
         self_ty: &rty::Ty,
-        trait_ref: &fhir::TraitRef,
+        trait_ref: &fhir::PolyTraitRef,
         kind: rty::ClosureKind,
         clauses: &mut Vec<rty::Clause>,
     ) -> QueryResult<()> {
-        let path = &trait_ref.path;
+        let path = &trait_ref.trait_ref;
+        let layer = Layer::list(self, trait_ref.bound_generic_params.len() as u32, &[], true);
+        env.push_layer(layer);
+
         let pred = rty::FnTraitPredicate {
             self_ty: self_ty.clone(),
             tupled_args: self.conv_ty(env, path.args[0].expect_type())?,
             output: self.conv_ty(env, &path.bindings[0].term)?,
             kind,
         };
-        clauses.push(rty::Clause::new(rty::ClauseKind::FnTrait(pred)));
+        let vars = trait_ref
```

Comment:
Can you please leave a comment here
```suggestion
        // FIXME(nlehmann) We should use `tcx.late_bound_vars` here instead of trusting our lowering
        let vars = trait_ref
```